libs/langchain/langchain/llms/nv_aiplay.py

```diff
@@ -0,0 +1,764 @@
+        kwargs["labels"] = kwargs.get("labels", self.labels)
+        kwargs["stop"] = stop if stop else getattr(self.client, "stop")
+        if kwargs.get("streaming", self.streaming) or kwargs["stop"]:
+            buffer = ""
+            for chunk in self._stream(messages, run_manager=run_manager, **kwargs):
+                buffer += chunk if isinstance(chunk, str) else chunk.text
+            responses = {"content": buffer}
+        else:
+            inputs = self.custom_preprocess(messages)
+            responses = self.get_generation(inputs, **kwargs)
+        outputs = self.custom_postprocess(responses)
+        return outputs
+
+    def _get_filled_chunk(
+        self, text: str, role: Optional[str] = "assistant"
+    ) -> Union[GenerationChunk, ChatGenerationChunk]:
+        """LLM and BasicChatModel have different streaming chunk specifications"""
+        if isinstance(self, LLM):
+            return GenerationChunk(text=text)
+        return ChatGenerationChunk(message=ChatMessageChunk(content=text, role=role))
+
+    def _stream(
+        self,
+        messages: Union[List[BaseMessage], str],
+        stop: Optional[List[str]] = None,
+        run_manager: Optional[CallbackManager] = None,
+        **kwargs: Any,
+    ) -> Iterator[Union[GenerationChunk, ChatGenerationChunk]]:
+        """Allows streaming to model!"""
+        inputs = self.custom_preprocess(messages)
+        kwargs["labels"] = kwargs.get("labels", self.labels)
+        kwargs["stop"] = stop if stop else getattr(self.client, "stop")
+        for response in self.get_stream(inputs, **kwargs):
+            chunk = self._get_filled_chunk(self.custom_postprocess(response))
+            yield chunk
+            if run_manager:
+                async_mtypes = (AsyncCallbackManager, AsyncCallbackManagerForLLMRun)
+                if isinstance(run_manager, async_mtypes):
+                    ## Edge case from LLM/SimpleChatModel default async methods
+                    asyncio.run(run_manager.on_llm_new_token(chunk.text, chunk=chunk))
+                else:
+                    run_manager.on_llm_new_token(chunk.text, chunk=chunk)
+
+    async def _astream(
+        self,
+        messages: Union[List[BaseMessage], str],
+        stop: Optional[List[str]] = None,
+        run_manager: Optional[AsyncCallbackManager] = None,
+        **kwargs: Any,
+    ) -> AsyncIterator[Union[GenerationChunk, ChatGenerationChunk]]:
+        inputs = self.custom_preprocess(messages)
+        kwargs["labels"] = kwargs.get("labels", self.labels)
+        kwargs["stop"] = stop if stop else getattr(self.client, "stop")
+        async for response in self.get_astream(inputs, **kwargs):
+            chunk = self._get_filled_chunk(self.custom_postprocess(response))
+            yield chunk
+            if run_manager:
+                await run_manager.on_llm_new_token(chunk.text, chunk=chunk)
+
+    def custom_preprocess(self, msgs: Union[str, Sequence]) -> List[Dict[str, str]]:
```

Comment:
@hinthornw Thanks! Yeah, the abstraction here settled in after a pretty annoying "follow-the-API" experience. At this point, the API for the existing models has luckily stabilized, but this is just a "centralize the components for maximum flexibility with minimal modification" effort. That way, a new model can come in with its own hard expectations (or make very weird new assumptions) and a single update was sufficient to account for it in every derivative instantiation.

On reflection, I think it would be fine to get rid of the string-to-message routine in favor of LLM positional arguments on the ContextBase/etc subclass. A similar solution could work with "system" positional arguments... but before I do that, I'd like to confirm that this would actually solve the issue you mentioned, or if there is some other considerations that need to be taken? As of right now, it does not look like the AI Playground team is planning to expose raw completion endpoints, but I would still prefer the flexibility for users to help them experiment with the playground in the context they're most comfortable with/is closest to their intended final stack. (Of note, there is a non-zero chance we might want to support vanilla completion or even more complex endpoint options in the future though, either with AI Playground or some more general efforts later on). 

What approach do you think would be the best to help support this kind of use case for now while the underlying API is still technically chat-only?

Reply:
Why need positional arguments here? 

If the underlying API is chat only, then we should just support a chat model. If there is a completion or other endpoint that is exposed at some point, we can support that through a separate class. It is much harder if we try to make a fake integration now that may or may not be compatible with the future API

Reply:
Honestly, fair enough. I'll take a look and see what I can do with the integrations.